enigma-swing/src/main/java/cuchaz/enigma/gui/panels/right/RightPanel.java

```diff
@@ -0,0 +1,75 @@
+package cuchaz.enigma.gui.panels.right;
+
+import cuchaz.enigma.gui.Gui;
+import cuchaz.enigma.utils.I18n;
+
+import javax.swing.JLabel;
+import javax.swing.JPanel;
+import javax.swing.JToggleButton;
+import java.awt.BorderLayout;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Supplier;
+
+public abstract class RightPanel extends JPanel {
+    public static final String DEFAULT = Type.STRUCTURE;
+    private static final Map<String, RightPanel> panels = new HashMap<>();
+
+    protected final JToggleButton button;
+	protected final JLabel title;
+	private final Supplier<String> titleProvider = () -> I18n.translate("right_panel." + this.getId() + ".title");
+
+    protected RightPanel(Gui gui) {
+        super(new BorderLayout());
+        this.button = new JToggleButton(titleProvider.get());
+        this.button.addActionListener(e -> gui.setRightPanel(this.getId(), true));
+		this.title = new JLabel(titleProvider.get());
+		this.add(this.title, BorderLayout.NORTH);
+    }
+
+    public abstract RightPanel.ButtonPosition getButtonPosition();
+
+    public abstract String getId();
+
+    @Override
+    public void setVisible(boolean visible) {
+        super.setVisible(visible);
+        this.getButton().setSelected(visible);
+    }
+
+    public void retranslateUi() {
+		String translatedTitle = this.titleProvider.get();
+        this.button.setText(translatedTitle);
+		this.title.setText(translatedTitle);
+    }
+
+    public JToggleButton getButton() {
+        return this.button;
+    }
+
+    public static void registerPanel(RightPanel panel) {
```

Comment:
`addPanel` would be better imo. Or rather, seeing that using each instance of a right panel needs a cast, using a `Class<?>` as key would be an alternative, so you would have this method to associate each panel class to an instance, and a generic method to retrieve a panel instance based on the given `Class<?>`, which would probably end up in a single unchecked cast warning

Reply:
>  and a generic method to retrieve a panel instance based on the given Class<?>
how would I write this?

Reply:
If you have a field `Map<Class<? extends RightPanel>, ? extends RightPanel> panels`, your method would be like this
```java
public static <T extends RightPanel> T getPanel(Class<T> clazz) {
    return (T) panels.get(clazz);
}
```

Reply:
You might want to throw an exception if the panel doesn't exist, instead of returning null